### PR TITLE
mm-768 - prepends the siteURL in the redirectConnect method

### LINKED ISSUE #768: /Jira Connect doesn't open mattermost's subpath url
We have our mattermost installation running in a subpath called /app/ (picked by us), when a user runs the /jira connect command, mattermost opens a new window that does not include /app/ in the url, but instead uses /plugins/jira/ directly.

Please make sure the URI includes the user's subpath as this is a configurable variable plugins need to be aware of.

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
#### Summary
using the `getState().entities.general.config.SiteURL` to fetch the site url and prepend it to the path as mentioned in the issue 
comments. 
Fixes #768 
#### Ticket Link
https://github.com/mattermost/mattermost-plugin-jira/issues/768

### CODE REVIEW

File: webapp/src/actions/index.ts [hunk 1]
```diff
@@ -412,8 +412,13 @@ export function handleConnectFlow(instanceID?: string) {
...
 export function redirectConnect(instanceID: string) {
     return async (dispatch, getState) => {
+        let baseUrl = getState().entities.general.config.SiteURL;
+
+        if (baseUrl && baseUrl[baseUrl.length - 1] === '/') {
+            baseUrl = baseUrl.substr(0, baseUrl.length - 1);
+        }
         const instancePrefix = '/instance/' + btoa(instanceID);
-        const target = '/plugins/' + PluginId + instancePrefix + '/user/connect';
+        const target = baseUrl + '/plugins/' + PluginId + instancePrefix + '/user/connect';
```
Comment (mickmister) [inline webapp/src/actions/index.ts:421]: Can we use this string interpolation syntax?

```suggestion
        const target = `${baseUrl}/plugins/${PluginId}${instancePrefix}/user/connect`;
```

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks @maisnamrajusingh :+1: I have one request to use template strings here

File: webapp/src/actions/index.ts (hunk 1 above)
Comment (maisnamraju) [reply webapp/src/actions/index.ts:421]: @mickmister  makes sense, i'm a fan of interpolation as well.

**Review by maisnamraju — APPROVED** [review]

**Comment by maisnamrajusingh** [comment]
@mickmister using string templates makes sense. I have committed it.

**Comment by maisnamraju** [comment]
@mickmister Please take a look and see if I missed anything. Cheers

**Review by mickmister — APPROVED** [review]

**Review by mickmister — CHANGES_REQUESTED** [review]
Sorry for asking for changes after approving, but can we please put the logic for calculating the site URL in a selector?

**Comment by maisnamrajusingh** [comment]
Mick, totally forgot about this. Updated this to move the code into selector.

File: webapp/src/actions/index.ts [hunk 2]
```diff
@@ -44,13 +56,15 @@ export const openCreateModal = (postId) => {
...
-});
+export const openCreateModalWithoutPost =
+  (description, channelId) => (dispatch) =>
+      dispatch({
+          type: ActionTypes.OPEN_CREATE_ISSUE_MODAL_WITHOUT_POST,
+          data: {
+              description,
+              channelId,
+          },
+      });
```
Comment (mickmister) [inline webapp/src/actions/index.ts:67]: Can we revert all of these styling changes not related to this PR?

**Review by mickmister — CHANGES_REQUESTED** [review]

File: webapp/src/actions/index.ts (hunk 2 above)
Comment (mickmister) [reply webapp/src/actions/index.ts:67]: Only parts of the code related to using the siteURL should be edited in this PR

**Comment by mickmister** [comment]
@maisnamrajusingh Can you check the rest of the repo's `webapp` folder to see if there other spots where we are not using the site URL when we should?

File: webapp/src/actions/index.ts (hunk 2 above)
Comment (sibasankarnayak) [reply webapp/src/actions/index.ts]: @mickmister  have removed all unwanted styles and checked only the changes to this PR is done .

**Comment by mickmister** [comment]
It looks like this repo already has a function to use the siteURL `getPluginServerRoute`. Can we use this selector instead of creating a new one?

https://github.com/mattermost/mattermost-plugin-jira/blob/d4f87fe8a91c1ef4b62228f0b8aeb4a14c6b1ca9/webapp/src/selectors/index.ts#L14

Also @maisnamrajusingh @sibasankarnayak can you please verify whether there are other spots that need to be updated that may need to use the site URL?

**Comment by sibasankarnayak** [comment]
> It looks like this repo already has a function to use the siteURL `getPluginServerRoute`. Can we use this selector instead of creating a new one?
> 
> https://github.com/mattermost/mattermost-plugin-jira/blob/d4f87fe8a91c1ef4b62228f0b8aeb4a14c6b1ca9/webapp/src/selectors/index.ts#L14
> 
> Also @maisnamrajusingh @sibasankarnayak can you please verify whether there are other spots that need to be updated that may need to use the site URL?

@mickmister these two are different  one returns `basePath + '/plugins/' + PluginId` and other one `baseUrl`
we can make a common function which will return basePath so we can use in common

**Comment by mickmister** [comment]
@sibasankarnayak The plugin never needs to calculate the base URL without `/plugins/' + PluginId`, so 1/5 we can get rid of the `getBaseUrl` function and just use the existing one.

**Comment by sibasankarnayak** [comment]
> @sibasankarnayak The plugin never needs to calculate the base URL without `/plugins/' + PluginId`, so 1/5 we can get rid of the `getBaseUrl` function and just use the existing one.

@mickmister  have removed the  new code and used the existing one , but can see there is a  high severity security vulnerability can you check this ?

File: webapp/src/selectors/index.ts [hunk 3]
```diff
@@ -35,8 +35,7 @@ export const getCurrentUserLocale = createSelector(
         }
 
         return locale;
-    }
-);
+    });
```
Comment (mickmister) [inline webapp/src/selectors/index.ts:38]: Can we revert this change?

**Review by mickmister — APPROVED** [review]
LGTM, just one request to revert a styling change

**Comment by mickmister** [comment]
> @mickmister have removed the new code and used the existing one , but can see there is a high severity security vulnerability can you check this ?

The vulnerability found is unrelated to this PR, and is not a case where we will change the code in this project

**Comment by mickmister** [comment]
@sibasankarnayak I replied to your comment above

File: webapp/src/selectors/index.ts (hunk 3 above)
Comment (sibasankarnayak) [reply webapp/src/selectors/index.ts]: Done

**Comment by dipak-demansol** [comment]
i need a matter-most that working on sub URL, already discussed this with dylan so he will provide me the access of that and after that i will test this.

**Review by dipak-demansol — COMMENTED** [review]
@maisnamrajusingh when i use this Command as **/jira instance connect URL** then i redirected on the New Tab/Page of jira **but i got the error,** Developer sir, are you able to re-generate this issue? 
SuPath issue 2021-11-19 23-34-44

File: webapp/src/selectors/index.ts [hunk 4]
```diff
@@ -37,7 +37,6 @@ export const getCurrentUserLocale = createSelector(
         return locale;
     }
 );
-
```
Comment (sibasankarnayak) [inline webapp/src/selectors/index.ts:40]: can remove unwanted change

Comment (sibasankarnayak) [inline webapp/src/selectors/index.ts:40]: can remove unwanted change

**Review by sibasankarnayak — APPROVED** [review]
LGTM!

just one change can remove the unwanted change

**Comment by maisnamrajusingh** [comment]
> 

@dipak-demansol  i didn't face this issue so far.

**Comment by dipak-demansol** [comment]
> > 
> 
> @dipak-demansol i didn't face this issue so far.

i shared you a server URL and My ID password, pls check with that.

**Comment by dipak-demansol** [comment]
> @maisnamrajusingh when i use this Command as **/jira instance connect URL** then i redirected on the New Tab/Page of jira **but i got the error,** Developer sir, are you able to re-generate this issue? SuPath issue 2021-11-19 23-34-44

@maisnamrajusingh Gentle reminder to Reply On this issue.

**Comment by maisnamrajusingh** [comment]
> 

@dipak-demansol just confirming that this is resolved and that this pr is ready for final review

**Comment by dipak-demansol** [comment]
> > @maisnamrajusingh when i use this Command as **/jira instance connect URL** then i redirected on the New Tab/Page of jira **but i got the error,** Developer sir, are you able to re-generate this issue? SuPath issue 2021-11-19 23-34-44
> 
> @maisnamrajusingh Gentle reminder to Reply On this issue.

@maisnamrajusingh i'm still facing this issue. let me know if is there any info or help you want from me to regenerate this issue.

**Review by dipak-demansol — COMMENTED** [review]
@maisnamraju https://github.com/mattermost/mattermost-plugin-jira/pull/785#issuecomment-998020748

**Comment by hanzei** [comment]
@maisnamrajusingh The git history of this PR is completely messed up. Would you please fix it?

**Comment by maisnamrajusingh** [comment]
@dipak-demansol can you give it a QA run, I have tested this on a cloud instance but don't have access to a sever instance so just want to make sure its working all across.

File: webapp/src/selectors/index.ts (hunk 4 above)
Comment (sibasankarnayak) [reply webapp/src/selectors/index.ts:40]: need to have a line gap

File: webapp/src/selectors/index.ts [hunk 5]
```diff
@@ -13,17 +13,19 @@ const getPluginState = (state) => state['plugins-' + PluginId] || {};
...
         basePath = new URL(config.SiteURL).pathname;
-
         if (basePath && basePath[basePath.length - 1] === '/') {
             basePath = basePath.substr(0, basePath.length - 1);
         }
-    }
+        const origin = new URL(config.SiteURL).origin;
 
-    return basePath + '/plugins/' + PluginId;
+        finalPath = `${origin}${basePath}/plugins/${PluginId}`;
```
Comment (maisnamrajusingh) [inline webapp/src/selectors/index.ts:26]: @hanzei  not sure if this is the right style you are looking for but it works.

File: server/user_server.go [hunk 6]
```diff
@@ -118,7 +118,7 @@ func (p *Plugin) httpOAuth1aComplete(w http.ResponseWriter, r *http.Request, ins
 	}{
 		JiraDisplayName:       juser.DisplayName + " (" + juser.Name + ")",
 		MattermostDisplayName: mmuser.GetDisplayName(model.ShowNicknameFullName),
-		RevokeURL:             path.Join(p.GetPluginURLPath(), instancePath(routeUserDisconnect, instance.GetID())),
+		RevokeURL:             p.GetSiteURL() + path.Join(p.GetPluginURLPath(), instancePath(routeUserDisconnect, instance.GetID())),
```
Comment (mickmister) [inline server/user_server.go:121]: Same here

**Review by mickmister — CHANGES_REQUESTED** [review]

File: webapp/src/selectors/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply webapp/src/selectors/index.ts:26]: yeah I tested this on a subpath. There was an issue with the urls inside the forms on connect which is fixed now.  WIll update the redirectConnect.

File: server/user_server.go (hunk 6 above)
Comment (maisnamrajusingh) [reply server/user_server.go:121]: certainly looks cleaner

File: server/user_cloud.go [hunk 7]
```diff
@@ -44,7 +44,7 @@ func (p *Plugin) httpACUserRedirect(w http.ResponseWriter, r *http.Request, inst
 		return respondErr(w, http.StatusBadRequest, err)
 	}
 
-	submitURL := path.Join(ci.Plugin.GetPluginURLPath(), instancePath(routeACUserConfirm, instanceID))
+	submitURL := p.GetSiteURL() + path.Join(ci.Plugin.GetPluginURLPath(), instancePath(routeACUserConfirm, instanceID))
```
[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply server/user_cloud.go]: Oops. should have taken a closer look at the methods available.

**Comment by maisnamrajusingh** [comment]
@mickmister  I have updated the logic but would like to get @dipak-demansol  to run through a QA run before I submit it to you for another review.

**Review by levb — CHANGES_REQUESTED** [review]

File: server/user_cloud.go [hunk 8]
```diff
@@ -44,7 +44,7 @@ func (p *Plugin) httpACUserRedirect(w http.ResponseWriter, r *http.Request, inst
 		return respondErr(w, http.StatusBadRequest, err)
 	}
 
-	submitURL := path.Join(ci.Plugin.GetPluginURLPath(), instancePath(routeACUserConfirm, instanceID))
+	submitURL := path.Join(p.GetPluginURL(), instancePath(routeACUserConfirm, instanceID))
```
[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply server/user_cloud.go:47]: @levb  I think I missed this one. I have done this else where.

[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply server/user_cloud.go]: @levb do you mean the following format `p.GetPluginURL() + instancePath(routeACUserConfirm, instanceID)`

**Comment by dipak-demansol** [comment]
Pls Ping Me When When all the Dev review is Done.

File: server/user_cloud.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (levb) [reply server/user_cloud.go]: See https://stackoverflow.com/questions/34668012/combine-url-paths-with-path-join, you can wrap it into a `func addURLPath(url, extraPath string) string`

[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply server/user_cloud.go]: I apologize, my knowledge on this is quite poor but ill go through the docs and update the PR for another review

[the remark this answers is not in the document]
Comment (maisnamrajusingh) [reply server/user_cloud.go]: @levb  please take a look now.

**Review by levb — APPROVED** [review]

File: webapp/src/actions/index.ts [hunk 9]
```diff
@@ -426,8 +426,11 @@ export function handleConnectFlow(instanceID?: string) {
 
 export function redirectConnect(instanceID: string) {
     return async (dispatch, getState) => {
+        const baseUrl = getState().entities.general.config.SiteURL;
+
         const instancePrefix = '/instance/' + btoa(instanceID);
-        const target = '/plugins/' + PluginId + instancePrefix + '/user/connect';
+        const target = baseUrl + '/plugins/' + PluginId + instancePrefix + '/user/connect';
```
Comment (mickmister) [inline webapp/src/actions/index.ts:432]: It looks like we're still manually computing the plugin URL here. It's concerning that my requests are not fulfilled verbatim, especially on such a small PR. Please refactor this to use `getPluginServerRoute`, so we know that all of the places that need to use the site URL are using it properly. Then please test using a server configured with a subpath.

File: webapp/src/selectors/index.ts (hunk 5 above)
Comment (maisnamrajusingh) [reply webapp/src/selectors/index.ts:27]: yeah base path starts with a slash. Come to think of it, I should remove this commit from this PR. The idea was to have the full path everywhere. Does it make sense to keep it or should we skip this out from this pr ?

File: webapp/src/actions/index.ts (hunk 9 above)
Comment (maisnamrajusingh) [reply webapp/src/actions/index.ts]: It's updated now. we are now using the `getPluginServerRoute` to generate the root url along with the plugin path.

**Review by dipak-demansol — APPROVED** [review]
**LGTM** Tested with normal server & subpath server and user is able to Connect the jira cloud on both server.

**Comment by maisnamrajusingh** [comment]
@mickmister 
https://github.com/mattermost/mattermost-plugin-jira/pull/785#discussion_r837663038
Regarding the issue i have already tested with the following test cases 
https://codesandbox.io/s/jest-test-forked-sg9j7w?file=/index.test.js 

I was hesitant to have the test written on this repo since there isn't a structure ready but if you want me to set up one. I'd be happy to

File: webapp/src/selectors/index.ts [hunk 10]
```diff
@@ -15,15 +15,18 @@ export const getPluginServerRoute = (state) => {
...
+    }
+    return finalPath;
```
Comment (mickmister) [inline webapp/src/selectors/index.ts:29]: This will not work when the site URL is not set, since `finalPath` is supposed to have the `plugins/(plugin id)` suffix, but it would be a blank string in this case. Although we generally check for the site URL to be defined on plugin startup, it's concerning that there is still a bug here, and potentially prone to other bugs if this code is modified later.

**Comment by mickmister** [comment]
@maisnamrajusingh Please re-request review if you need someone to re-review a PR

**Review by mickmister — CHANGES_REQUESTED** [review]
@maisnamrajusingh Can you please revert the whitespace change in `webpack.config.js`, in order to keep the commit history for the file clean? Otherwise LGTM

**Comment by maisnamrajusingh** [comment]
> 

will do.

File: webapp/src/selectors/index.ts (hunk 10 above)
Comment (maisnamrajusingh) [reply webapp/src/selectors/index.ts:29]: @mickmister I see two options here 

- we fetch the site url from window.location.href and apply instead of config.SiteURL
- throw an error when the site url is not configured. 

I like the first option. what do you suggest

Comment (maisnamrajusingh) [reply webapp/src/selectors/index.ts]: [Done Michael. I have added a handler that checks the url is present or not ](https://github.com/mattermost/mattermost-plugin-jira/blob/4acda125c42617a33d78bf741bd958273e2d5b88/webapp/src/selectors/index.ts#L14-L31)

Comment (mickmister) [reply webapp/src/selectors/index.ts]: We don't want to use the `window.locatiom.href`'s `pathname ` though. That will contain `/channels/(team)` etc.
